python/xoscar/collective/tests/test_cupy.py

```diff
@@ -0,0 +1,313 @@
+        recvbuf, cupy.array([[1, 2, 3], [1, 2, 3]], dtype=cupy.float32)
+    )
+
+
+def worker_scatter(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+
+    sendbuf = cupy.array([1, 2, 3, 4, 5, 6], dtype=cupy.float32)
+    recvbuf = cupy.array([0, 0, 0], dtype=cupy.float32)
+    root = 0
+
+    xc.scatter(context, sendbuf, recvbuf, root)
+
+    if rank == 0:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([1.0, 2.0, 3.0]))
+    elif rank == 1:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([4.0, 5.0, 6.0]))
+
+
+def worker_gather(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+
+    if rank == 0:
+        sendbuf = cupy.array([1, 2, 3], dtype=cupy.float32)
+    else:
+        sendbuf = cupy.array([4, 5, 6], dtype=cupy.float32)
+
+    recvbuf = cupy.array([0, 0, 0, 0, 0, 0], dtype=cupy.float32)
+    root = 0
+
+    xc.gather(context, sendbuf, recvbuf, root)
+
+    if rank == 0:
+        cupy.testing.assert_array_equal(
+            recvbuf, cupy.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0])
+        )
+
+
+def worker_reduce_scatter(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+
+    sendbuf = cupy.array([1, 2, 3, 4], dtype=cupy.float32)
+    recvbuf = cupy.array([0, 0], dtype=cupy.float32)
+    op = xc.ReduceOp.SUM
+
+    xc.reduce_scatter(context, sendbuf, recvbuf, op)
+
+    if rank == 0:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([2.0, 4.0]))
+    elif rank == 1:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([6.0, 8.0]))
+
+
+def worker_all_to_all(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+
+    if rank == 0:
+        sendbuf = cupy.array([[1, 2], [3, 4]], dtype=cupy.float32)
+    else:
+        sendbuf = cupy.array([[2, 4], [6, 8]], dtype=cupy.float32)
+    recvbuf = cupy.zeros((2, 2), dtype=cupy.float32)
+
+    xc.all_to_all(context, sendbuf, recvbuf)
+
+    if rank == 0:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([[1.0, 2.0], [2.0, 4.0]]))
+    elif rank == 1:
+        cupy.testing.assert_array_equal(recvbuf, cupy.array([[3.0, 4.0], [6.0, 8.0]]))
+
+
+def worker_barrier(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+    sendbuf = cupy.array([[1, 2, 3], [1, 2, 3]], dtype=cupy.float32)
+    recvbuf = cupy.zeros_like(sendbuf, dtype=cupy.float32)
+    op = xc.ReduceOp.SUM
+
+    xc.allreduce(context, sendbuf, recvbuf, op)
+    xc.barrier(context)
+
+    cupy.testing.assert_array_equal(recvbuf, cupy.array(sendbuf * 2))
+
+
+def test_scatter_buffer(rank, nccl_id, device_id):
```

Comment:
There will be an error that `fixture 'rank' not found`, maybe `test_scatter_buffer` should be `worker_scatter_buffer`